### PR TITLE
Added restrictions to avoid desync problems

### PR DESCRIPTION
**Pull request by lsebrie**
Minor changes on classes to avoid thread syncing issues

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 181 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 75 major
* ![MINOR][MINOR] 92 minor
* ![INFO][INFO] 12 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [IndexedBlockStore.java#L420](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java#L420): Make the "java.util.List" class "Serializable" or don't write it. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2118)
1. ![CRITICAL][CRITICAL] [IndexedBlockStore.java#L438](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java#L438): Object deserialization is used in org.ethereum.db.IndexedBlockStore$1.deserialize(DataInput, int) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3AOBJECT_DESERIALIZATION)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L117](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L117): Constructor new co.rsk.core.bc.BlockChainImpl(Repository, BlockStore, ReceiptStore, PendingState, EthereumListener, AdminInfo, BlockValidator) makes call to non-final method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APCOA_PARTIALLY_CONSTRUCTED_OBJECT_ACCESS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L199](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L199): The Cyclomatic Complexity of this method "internalTryToConnect" is 22 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L201](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L201): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L308](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L308): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L364](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L364): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L375](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L375): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L380](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L380): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L469](https://github.com/rsksmart/rskj/blob/a04bfb3a8443affca726a08674403f84dd6807e7/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L469): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 181 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 77 major
* ![MINOR][MINOR] 92 minor
* ![INFO][INFO] 10 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [IndexedBlockStore.java#L420](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java#L420): Make the "java.util.List" class "Serializable" or don't write it. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2118)
1. ![CRITICAL][CRITICAL] [IndexedBlockStore.java#L438](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java#L438): Object deserialization is used in org.ethereum.db.IndexedBlockStore$1.deserialize(DataInput, int) [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3AOBJECT_DESERIALIZATION)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L101](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L101): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L118](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L118): Constructor new co.rsk.core.bc.BlockChainImpl(Repository, BlockStore, ReceiptStore, PendingState, EthereumListener, AdminInfo, BlockValidator, RskSystemProperties) makes call to non-final method [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APCOA_PARTIALLY_CONSTRUCTED_OBJECT_ACCESS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L200](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L200): The Cyclomatic Complexity of this method "internalTryToConnect" is 22 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L202](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L202): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L309](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L309): Method co.rsk.core.bc.BlockChainImpl.internalTryToConnect(Block) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L365](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L365): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L376](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L376): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [BlockChainImpl.java#L381](https://github.com/rsksmart/rskj/blob/bd8c4bb55bfd2c4ad88a43552e3a71b83a15dc15/rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java#L381): Return an empty collection instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java [hunk 1]
```diff
@@ -563,7 +563,7 @@ private void flushData() {
         return Block.getTxTrie(transactions).getHash();
     }
 
-    public static byte[] calcReceiptsTrie(List<TransactionReceipt> receipts) {
+    public synchronized static byte[] calcReceiptsTrie(List<TransactionReceipt> receipts) {
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java]: Why add synchronized to a static method? which is the use case? The collection argument could mutate EVEN with the synchronized added to method. The variables are locals, and the trie is immutable

Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java]: It avoids writing multiple times on Trie hashes property by calling methods getHash( ) -> toMessage( ) -> getHash(K) -> setHash( )
Maybe we can get to a better solution for this.

Comment (ajlopezrsk) [reply rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java]: I don't still see the use case. The Trie hash property is only calculated once, at the en do f the method. I don't understand your reference to "-> toMessage" The above static method is only used in miner code. 

Out of scope: there is another private method in BlockResult. We should unify both. But it is out of scope of this pull request.

Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/core/bc/BlockChainImpl.java]: I was confused about the static method, a new Trie is created everytime on the static method so there is no problem with the getHash method inside of calcReceiptsTrie. I'll remove the synchronized on this method. 

If getHash is called on another place with an already created Trie this problem could arise, if it's not precalculated to be stored inside of Trie it should use some kind of synchronization.

**Pull request merged**